transports/webrtc-websys/src/stream.rs

```diff
@@ -0,0 +1,268 @@
+                peer_connection
+                    .create_data_channel_with_data_channel_dict(LABEL, &data_channel_dict)
+            }
+            false => peer_connection.create_data_channel(LABEL),
+        };
+        dc.set_binary_type(ARRAY_BUFFER_BINARY_TYPE); // Hardcoded here, it's the only type we use
+        dc
+    }
+}
+
+/// Substream over the Connection
+pub struct Stream {
+    inner: SendWrapper<StreamInner>,
+    state: State,
+    read_buffer: Bytes,
+}
+
+impl Stream {
+    /// Create a new WebRTC Substream
+    pub fn new(channel: RtcDataChannel) -> Self {
+        Self {
+            inner: SendWrapper::new(StreamInner::new(channel)),
+            read_buffer: Bytes::new(),
+            state: State::Open,
+        }
+    }
+
+    /// Return the underlying RtcDataChannel
+    pub fn channel(&self) -> &RtcDataChannel {
+        self.inner.io.as_ref()
+    }
+}
+
+/// Inner Stream to make Sendable
+struct StreamInner {
+    io: FramedDc,
+}
+
+impl StreamInner {
+    pub fn new(channel: RtcDataChannel) -> Self {
+        Self {
+            io: framed_dc::new(channel),
+        }
+    }
+}
+
+impl AsyncRead for Stream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        loop {
+            self.state.read_barrier()?;
+
+            // Return buffered data, if any
+            if !self.read_buffer.is_empty() {
+                let n = std::cmp::min(self.read_buffer.len(), buf.len());
+                let data = self.read_buffer.split_to(n);
+                buf[0..n].copy_from_slice(&data[..]);
+
+                return Poll::Ready(Ok(n));
+            }
+
+            let Self {
+                read_buffer,
+                state,
+                inner,
+                ..
+            } = &mut *self;
+
+            match ready!(io_poll_next(&mut inner.io, cx))? {
+                Some((flag, message)) => {
+                    if let Some(flag) = flag {
+                        state.handle_inbound_flag(flag, read_buffer);
+                    }
+
+                    debug_assert!(read_buffer.is_empty());
+                    if let Some(message) = message {
+                        *read_buffer = message.into();
+                    }
+                    // continue to loop
+                }
+                None => {
+                    state.handle_inbound_flag(Flag::FIN, read_buffer);
+                    return Poll::Ready(Ok(0));
+                }
+            }
+        }
+    }
+}
+
+impl AsyncWrite for Stream {
+    /// Attempt to write bytes from buf into the object.
+    /// On success, returns Poll::Ready(Ok(num_bytes_written)).
+    /// If the object is not ready for writing,
+    /// the method returns Poll::Pending and
+    /// arranges for the current task (via cx.waker().wake_by_ref())
+    /// to receive a notification when the object becomes writable
+    /// or is closed.
+    ///
```

Comment:
```suggestion
```

This is not specific to this trait implementation but applies to all implementations of the trait, right?

Reply:
Right! This was a note to myself earlier as I was trying to learn how the signal flow worked.

I will remove now.